### ISSUE TITLE
Compilation errors with new CoDiPack version

### ISSUE THREAD

**Issue by jacksonvanover** [LLNL/adapt-fp#5]
The latest versions of CoDiPack and ADAPT do not appear to work, i.e., I encounter errors like the one below when compiling any of the provided adapt examples. I think it has to do with the latest version of CoDiPack as I encounter the same error when running `make` in the CoDiPack directory. The issue goes away when reverting to CoDiPack version 1.9.3 which then requires using the version of ADAPT at commit be4cda8.

```
g++ -g -O0 -Wall --std=c++11 -I/g/g15/vanover1/CoDiPack/include -I../ -DCODI_ReversalZeroesAdjoints=0 -o sanity-adapt sanity-adapt.cpp
In file included from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/../../expressions/lhsExpressionInterface.hpp:42:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/externalFunctionTapeInterface.hpp:39,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/fullTapeInterface.hpp:41,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/activeType.hpp:39,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:38,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/../../expressions/expressionInterface.hpp:105:22: error: specialization of 'template<class T_Type, class> struct codi::RealTraits::TraitsImplementation' in different namespace [-fpermissive]
   struct RealTraits::TraitsImplementation<T_Type, ExpressionTraits::EnableIfExpression<T_Type>> {
                      ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/dataManagementTapeInterface.hpp:40:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/fullTapeInterface.hpp:40,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/activeType.hpp:39,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:38,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/../../traits/realTraits.hpp:60:12: error:   from definition of 'template<class T_Type, class> struct codi::RealTraits::TraitsImplementation' [-fpermissive]
     struct TraitsImplementation {
            ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/externalFunctionTapeInterface.hpp:39:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/fullTapeInterface.hpp:41,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/activeType.hpp:39,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:38,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/../../expressions/lhsExpressionInterface.hpp:203:22: error: specialization of 'template<class T_Type, class> struct codi::RealTraits::DataExtraction' in different namespace [-fpermissive]
   struct RealTraits::DataExtraction<T_Type, ExpressionTraits::EnableIfLhsExpression<T_Type>> {
                      ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/dataManagementTapeInterface.hpp:40:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/fullTapeInterface.hpp:40,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/activeType.hpp:39,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:38,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/../../traits/realTraits.hpp:158:12: error:   from definition of 'template<class T_Type, class> struct codi::RealTraits::DataExtraction' [-fpermissive]
     struct DataExtraction {
            ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/externalFunctionTapeInterface.hpp:39:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/fullTapeInterface.hpp:41,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/activeType.hpp:39,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:38,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/../../expressions/lhsExpressionInterface.hpp:229:22: error: specialization of 'template<class T_Type, class> struct codi::RealTraits::TapeRegistration' in different namespace [-fpermissive]
   struct RealTraits::TapeRegistration<T_Type, ExpressionTraits::EnableIfLhsExpression<T_Type>> {
                      ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/dataManagementTapeInterface.hpp:40:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/fullTapeInterface.hpp:40,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/activeType.hpp:39,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:38,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/../../traits/realTraits.hpp:189:12: error:   from definition of 'template<class T_Type, class> struct codi::RealTraits::TapeRegistration' [-fpermissive]
     struct TapeRegistration {
            ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/jacobianBaseTape.hpp:50:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/tapeTraits.hpp:41,
                 from /g/g15/vanover1/CoDiPack/include/codi/tapes/forwardEvaluation.hpp:43,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:45,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/../traits/computationTraits.hpp:186:29: error: specialization of 'template<class T_Outer, class T_Inner, class> struct codi::ComputationTraits::AdjointConversionImpl' in different namespace [-fpermissive]
   struct ComputationTraits::AdjointConversionImpl<Type, Type> {
                             ^
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/../traits/computationTraits.hpp:61:12: error:   from definition of 'template<class T_Outer, class T_Inner, class> struct codi::ComputationTraits::AdjointConversionImpl' [-fpermissive]
     struct AdjointConversionImpl {
            ^
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/../traits/computationTraits.hpp:200:29: error: specialization of 'template<class T_Outer, class T_Inner, class> struct codi::ComputationTraits::AdjointConversionImpl' in different namespace [-fpermissive]
   struct ComputationTraits::AdjointConversionImpl<
                             ^
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/../traits/computationTraits.hpp:61:12: error:   from definition of 'template<class T_Outer, class T_Inner, class> struct codi::ComputationTraits::AdjointConversionImpl' [-fpermissive]
     struct AdjointConversionImpl {
            ^
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/../traits/computationTraits.hpp:221:29: error: specialization of 'template<class T_Outer, class T_Inner, class> struct codi::ComputationTraits::AdjointConversionImpl' in different namespace [-fpermissive]
   struct ComputationTraits::AdjointConversionImpl<Type, std::complex<Type>> {
                             ^
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/../traits/computationTraits.hpp:61:12: error:   from definition of 'template<class T_Outer, class T_Inner, class> struct codi::ComputationTraits::AdjointConversionImpl' [-fpermissive]
     struct AdjointConversionImpl {
            ^
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/../traits/computationTraits.hpp:235:29: error: specialization of 'template<class T_Jacobian, class> struct codi::ComputationTraits::TransposeImpl' in different namespace [-fpermissive]
   struct ComputationTraits::TransposeImpl<T, typename std::enable_if<std::is_floating_point<T>::value>::type> {
                             ^
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/../traits/computationTraits.hpp:91:12: error:   from definition of 'template<class T_Jacobian, class> struct codi::ComputationTraits::TransposeImpl' [-fpermissive]
     struct TransposeImpl {
            ^
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/../traits/computationTraits.hpp:247:29: error: specialization of 'template<class T_Jacobian, class> struct codi::ComputationTraits::TransposeImpl' in different namespace [-fpermissive]
   struct ComputationTraits::TransposeImpl<std::complex<Inner>> {
                             ^
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/../traits/computationTraits.hpp:91:12: error:   from definition of 'template<class T_Jacobian, class> struct codi::ComputationTraits::TransposeImpl' [-fpermissive]
     struct TransposeImpl {
            ^
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/../traits/computationTraits.hpp:261:29: error: specialization of 'template<class T_Lhs, class T_Rhs, class> struct codi::ComputationTraits::UpdateImpl' in different namespace [-fpermissive]
   struct ComputationTraits::UpdateImpl<Inner, std::complex<Inner>> {
                             ^
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/../traits/computationTraits.hpp:118:12: error:   from definition of 'template<class T_Lhs, class T_Rhs, class> struct codi::ComputationTraits::UpdateImpl' [-fpermissive]
     struct UpdateImpl {
            ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/misc/adjointVectorAccess.hpp:41:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/jacobianBaseTape.hpp:52,
                 from /g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/tapeTraits.hpp:41,
                 from /g/g15/vanover1/CoDiPack/include/codi/tapes/forwardEvaluation.hpp:43,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:45,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/misc/../../tools/data/direction.hpp:289:22: error: specialization of 'template<class T_Type, class> struct codi::RealTraits::IsTotalZero' in different namespace [-fpermissive]
   struct RealTraits::IsTotalZero<T_Type, GradientTraits::EnableIfDirection<T_Type>> {
                      ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/dataManagementTapeInterface.hpp:40:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/fullTapeInterface.hpp:40,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/activeType.hpp:39,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:38,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/../../traits/realTraits.hpp:100:12: error:   from definition of 'template<class T_Type, class> struct codi::RealTraits::IsTotalZero' [-fpermissive]
     struct IsTotalZero {
            ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/misc/adjointVectorAccess.hpp:41:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/jacobianBaseTape.hpp:52,
                 from /g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/tapeTraits.hpp:41,
                 from /g/g15/vanover1/CoDiPack/include/codi/tapes/forwardEvaluation.hpp:43,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:45,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/misc/../../tools/data/direction.hpp:306:22: error: specialization of 'template<class T_Type, class> struct codi::RealTraits::IsTotalFinite' in different namespace [-fpermissive]
   struct RealTraits::IsTotalFinite<T_Type, GradientTraits::EnableIfDirection<T_Type>> {
                      ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/dataManagementTapeInterface.hpp:40:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/fullTapeInterface.hpp:40,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/activeType.hpp:39,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:38,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/../../traits/realTraits.hpp:82:12: error:   from definition of 'template<class T_Type, class> struct codi::RealTraits::IsTotalFinite' [-fpermissive]
     struct IsTotalFinite {
            ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/tapeTraits.hpp:42:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/tapes/forwardEvaluation.hpp:43,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:45,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/primalValueBaseTape.hpp:1255:28: error: specialization of 'template<class Expr> struct codi::ExpressionTraits::NumberOfActiveTypeArguments' in different namespace [-fpermissive]
   struct ExpressionTraits::NumberOfActiveTypeArguments<JacobianExpression<size>> {
                            ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/../../traits/realTraits.hpp:42:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/dataManagementTapeInterface.hpp:40,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/fullTapeInterface.hpp:40,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/activeType.hpp:39,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:38,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/../../traits/expressionTraits.hpp:184:12: error:   from definition of 'template<class Expr> struct codi::ExpressionTraits::NumberOfActiveTypeArguments' [-fpermissive]
     struct NumberOfActiveTypeArguments : public CompileTimeTraversalLogic<size_t, NumberOfActiveTypeArguments<Expr>> {
            ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/tapeTraits.hpp:42:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/tapes/forwardEvaluation.hpp:43,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:45,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/tapes/../traits/../tapes/primalValueBaseTape.hpp:1261:28: error: specialization of 'template<class Expr> struct codi::ExpressionTraits::NumberOfConstantTypeArguments' in different namespace [-fpermissive]
   struct ExpressionTraits::NumberOfConstantTypeArguments<JacobianExpression<size>> {
                            ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/../../traits/realTraits.hpp:42:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/dataManagementTapeInterface.hpp:40,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/fullTapeInterface.hpp:40,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/activeType.hpp:39,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:38,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/../../traits/expressionTraits.hpp:206:12: error:   from definition of 'template<class Expr> struct codi::ExpressionTraits::NumberOfConstantTypeArguments' [-fpermissive]
     struct NumberOfConstantTypeArguments
            ^
In file included from /g/g15/vanover1/CoDiPack/include/codi.hpp:45:0,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/tapes/forwardEvaluation.hpp:191:22: error: specialization of 'template<class T_Type, class> struct codi::RealTraits::IsTotalFinite' in different namespace [-fpermissive]
   struct RealTraits::IsTotalFinite<T_Type, TapeTraits::EnableIfForwardTape<typename T_Type::Tape>> {
                      ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/dataManagementTapeInterface.hpp:40:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/fullTapeInterface.hpp:40,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/activeType.hpp:39,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:38,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/../../traits/realTraits.hpp:82:12: error:   from definition of 'template<class T_Type, class> struct codi::RealTraits::IsTotalFinite' [-fpermissive]
     struct IsTotalFinite {
            ^
In file included from /g/g15/vanover1/CoDiPack/include/codi.hpp:45:0,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/tapes/forwardEvaluation.hpp:207:22: error: specialization of 'template<class T_Type, class> struct codi::RealTraits::IsTotalZero' in different namespace [-fpermissive]
   struct RealTraits::IsTotalZero<T_Type, TapeTraits::EnableIfForwardTape<typename T_Type::Tape>> {
                      ^
In file included from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/dataManagementTapeInterface.hpp:40:0,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/fullTapeInterface.hpp:40,
                 from /g/g15/vanover1/CoDiPack/include/codi/expressions/activeType.hpp:39,
                 from /g/g15/vanover1/CoDiPack/include/codi.hpp:38,
                 from ../adapt.h:50,
                 from sanity-adapt.cpp:2:
/g/g15/vanover1/CoDiPack/include/codi/expressions/../tapes/interfaces/../../traits/realTraits.hpp:100:12: error:   from definition of 'template<class T_Type, class> struct codi::RealTraits::IsTotalZero' [-fpermissive]
     struct IsTotalZero {
            ^
make: *** [sanity-adapt] Error 1
```

**Comment by lam2mo** [comment]
I have not been able to reproduce these build failures. I just recompiled everything using CoDiPack 2.0.2 and the latest commit of ADAPT and got no error (or even warnings) with GCC 7.4.0 on RHEL 7.9. Which compiler are you using?

**Comment by lam2mo** [comment]
I did find these pages that imply it may be a GCC bug in older versions:
- https://stackoverflow.com/questions/25594644/warning-specialization-of-template-in-different-namespace
- https://gcc.gnu.org/bugzilla/show_bug.cgi?id=56480

**Comment by jacksonvanover** [comment]
> I did find these pages that imply it may be a GCC bug in older versions:
> 
>     * https://stackoverflow.com/questions/25594644/warning-specialization-of-template-in-different-namespace
> 
>     * https://gcc.gnu.org/bugzilla/show_bug.cgi?id=56480

Solved! I was using gcc v4.9.3

Compilation errors also occur for gcc v6.1.0 but upgrading to 7.1.0 fixes it. Thanks!
